jsv2/kv/bucket.go

```diff
@@ -0,0 +1,398 @@
+		Values() uint64
+
+		// History returns the configured history kept per key
+		History() int64
+
+		// TTL is how long the bucket keeps values for
+		TTL() time.Duration
+
+		// BackingStore indicates what technology is used for storage of the bucket
+		BackingStore() string
+	}
+
+	// KeyWatcher is what is returned when doing a watch.
+	KeyWatcher interface {
+		// Context returns watcher context optionally provided by nats.Context option.
+		Context() context.Context
+		// Updates returns a channel to read any updates to entries.
+		Updates() <-chan KeyValueEntry
+		// Stop will stop this watcher.
+		Stop() error
+	}
+
+	WatchOpt interface {
+		configureWatcher(opts *watchOpts) error
+	}
+
+	watchOpts struct {
+		// Do not send delete markers to the update channel.
+		ignoreDeletes bool
+		// Include all history per subject, not just last one.
+		includeHistory bool
+		// retrieve only the meta data of the entry
+		metaOnly bool
+	}
+
+	watchOptFn func(opts *watchOpts) error
+
+	// KeyValueBucketStatus represents status of a Bucket, implements KeyValueStatus
+	KeyValueBucketStatus struct {
+		nfo    *jetstream.StreamInfo
+		bucket string
+	}
+)
+
+func (e *kve) Bucket() string        { return e.bucket }
+func (e *kve) Key() string           { return e.key }
+func (e *kve) Value() []byte         { return e.value }
+func (e *kve) Revision() uint64      { return e.revision }
+func (e *kve) Created() time.Time    { return e.created }
+func (e *kve) Delta() uint64         { return e.delta }
+func (e *kve) Operation() KeyValueOp { return e.op }
+
+func keyValid(key string) bool {
+	if len(key) == 0 || key[0] == '.' || key[len(key)-1] == '.' {
+		return false
+	}
+	return validKeyRe.MatchString(key)
+}
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(ctx context.Context, key string) (KeyValueEntry, error) {
+	e, err := kv.get(ctx, key, kvLatestRevision)
+	if err != nil {
+		if err == ErrKeyDeleted {
+			return nil, ErrKeyNotFound
+		}
+		return nil, err
+	}
+
+	return e, nil
+}
+
+// GetRevision returns a specific revision value for the key.
+func (kv *kvs) GetRevision(ctx context.Context, key string, revision uint64) (KeyValueEntry, error) {
+	e, err := kv.get(ctx, key, revision)
+	if err != nil {
+		if err == ErrKeyDeleted {
+			return nil, ErrKeyNotFound
+		}
+		return nil, err
+	}
+
+	return e, nil
+}
+
+func (kv *kvs) get(ctx context.Context, key string, revision uint64) (KeyValueEntry, error) {
+	if !keyValid(key) {
+		return nil, ErrInvalidKey
+	}
+
+	var b strings.Builder
+	b.WriteString(kv.pre)
+	b.WriteString(key)
+
+	var m *jetstream.RawStreamMsg
+	var err error
+	if revision == kvLatestRevision {
+		m, err = kv.stream.GetLastMsgForSubject(ctx, b.String())
+	} else {
+		m, err = kv.stream.GetMsg(ctx, revision)
+		if err == nil && m.Subject != b.String() {
+			return nil, ErrKeyNotFound
+		}
+	}
+
+	if err != nil {
+		if err == jetstream.ErrMsgNotFound {
+			err = ErrKeyNotFound
```

Comment:
We often do this for some errors:
```
if err == jetstream.ErrMsgNotFound {
			err = ErrKeyNotFound
```
I think maybe a sign that error should have ben wrapped instead?  Think it would help have `errors.Is(err, ErrKeyNotFound)` but then be able to distinguish from reason being that it was because key was deleted, got 404 from API or because revision number value does not match the retrieved so transformed into key not found error (like a few lines above).

Reply:
Makes a ton of sense. I'll scan the `jsv2` codebase and see where it makes sense. BTW, the errors implementation in this PR will change quite a bit, as right now I don't yet have a common type to represent both client and API errors (like we have in `js.go`).